Remove use of `future` in `band_to_tridiag` and helpers

msimberg opened the pull request:
Not a finished PR and can be split into multiple PRs.

- Makes copies high priority in `band_to_tridiag`
- Removes the use of `make_future` from `Pipeline`
- Removes the use of `make_future` in `band_to_tridiag` (use `any_sender` instead)
- ~Adds annotations to `band_to_tridiag` (to be removed)~
- ~Adds a way to give a scheduling hint to `Policy` (may be removed, unsure if it helps enough; also needs to not be applied to the `Backend::GPU` scheduler).~

Also needs general cleanup. This is mainly meant for testing and benchmarks. Requires this fix in pika: https://github.com/pika-org/pika/pull/203.

To do:

- [x] fix compilation of `test_pipeline`
- [x] remove annotations
- [x] possibly remove hint in `Policy`

msimberg commented:
bors try

msimberg commented on include/dlaf/communication/kernels/all_reduce.h:
```diff
@@ -63,8 +63,8 @@ auto allReduceInPlace(common::PromiseGuard<comm::Communicator> pcomm, MPI_Op red
 
-template <class T>
-void scheduleAllReduce(pika::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
+template <class T, class CommSender>
```
Put `CommSender` first, unless `T` has to be specified explicitly.

msimberg commented on include/dlaf/communication/kernels/all_reduce.h:
```diff
@@ -89,24 +89,23 @@ void scheduleAllReduce(pika::future<common::PromiseGuard<comm::Communicator>> pc
 }
 
-template <class T>
+template <class T, class CommSender>
```
Same here.

msimberg replied:
Not here.

msimberg commented on include/dlaf/communication/kernels/reduce.h:
```diff
@@ -58,9 +58,8 @@ void reduceSend(comm::IndexT_MPI rank_root, common::PromiseGuard<comm::Communica
-template <class Sender>
-auto senderReduceRecvInPlace(pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
-                             MPI_Op reduce_op, Sender&& tile) {
+template <class Sender, class CommSender>
```
`CommSender` first.

msimberg commented on include/dlaf/communication/kernels/reduce.h:
```diff
@@ -97,10 +97,8 @@ auto senderReduceRecvInPlace(pika::future<common::PromiseGuard<comm::Communicato
-auto senderReduceSend(comm::IndexT_MPI rank_root,
-                      pika::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
-                      Sender&& tile) {
+template <class Sender, class CommSender>
```
`CommSender` first.

msimberg commented on include/dlaf/communication/kernels/reduce.h:
```diff
@@ -129,9 +127,9 @@ auto senderReduceSend(comm::IndexT_MPI rank_root,
-template <class T>
-void scheduleReduceRecvInPlace(pika::future<common::PromiseGuard<comm::Communicator>> pcomm,
-                               MPI_Op reduce_op, pika::future<matrix::Tile<T, Device::CPU>> tile) {
+template <class T, class CommSender>
```
`CommSender` first.

msimberg commented on include/dlaf/communication/kernels/reduce.h:
```diff
@@ -158,37 +157,37 @@ void scheduleReduceRecvInPlace(pika::future<common::PromiseGuard<comm::Communica
-template <class T>
-void scheduleReduceSend(comm::IndexT_MPI rank_root,
-                        pika::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
+template <class T, class CommSender>
```
`CommSender` first.

msimberg commented on include/dlaf/communication/kernels/reduce.h:
```diff
@@ -198,15 +197,15 @@ void scheduleReduceSend(comm::IndexT_MPI rank_root,
-template <class T, Device D>
-void scheduleReduceSend(comm::IndexT_MPI rank_root,
-                        pika::future<common::PromiseGuard<comm::Communicator>> pcomm, MPI_Op reduce_op,
+template <class T, Device D, class CommSender>
```
`CommSender` first.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -347,14 +347,15 @@ class SweepWorker {
 template <Device D, class T>
 TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
     const SizeType b, Matrix<const T, D>& mat_a) noexcept {
+  pika::scoped_annotation ann1("band_to_tridiag");
```
Remove.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -393,43 +394,45 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
-      for (int i = 0; i < nb / b - 1; ++i) {
-        deps.push_back(sf);
+  {
+    pika::scoped_annotation ann2("band_to_tridiag/copy");
```
Remove.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -448,42 +451,64 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
+          dlaf::internal::transformDetach(policy, copy_tridiag_task);
+    }
+    else {
+      ex::start_detached(std::forward<decltype(dep)>(dep));
```
Not all input senders need to be started (and in that case this is an unnecessary allocation). Does that ever happen?

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -448,42 +451,64 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
-    for (SizeType step = 0; step < steps; ++step) {
-      auto dep_index = std::min(step + 1, deps.size() - 1);
+  {
+    pika::scoped_annotation ann3("band_to_tridiag/sweeps");
```
Remove.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -448,42 +451,64 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
+    for (SizeType sweep = 0; sweep < sweeps; ++sweep) {
+      // Create the first max_workers workers and then reuse them.
+      auto& w_pipeline = workers[sweep % max_workers];
+      auto thread_hint = static_cast<std::remove_cv_t<decltype(num_threads)>>(sweep) % num_threads;
+      dlaf::internal::Policy<Backend::MC>
+          policy(pika::threads::thread_priority::high,
+                 pika::threads::thread_schedule_hint(pika::threads::thread_schedule_hint_mode::thread,
+                                                     thread_hint));
```
Does this actually help? Non-exhaustive tests seem to say yes, but not significantly.

msimberg replied:
Removed.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -448,42 +451,64 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
-          ex::then(store_tau_v) | ex::start_detached();
-      deps[step] = dlaf::internal::whenAllLift(w_pipeline(), deps[dep_index]) |
-                   dlaf::internal::transform(policy_hp, cont_sweep) | ex::make_future();
+      if (thread_hint == 0) {
+        pika::this_thread::yield();
+      }
```
Definitely remove.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -448,42 +451,64 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
   if (!isComplex_v<T>) {
     // only needed for real types as they don't perform sweep size-2
-    copy_tridiag(size - 2, deps[0]);
+    pika::scoped_annotation ann4("band_to_tridiag/non-complex-copy_tridiag");
```
Remove.

msimberg commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -448,42 +451,64 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
+  }
+
+  {
+    pika::scoped_annotation ann4("band_to_tridiag/final-copy_tridiag");
```
Remove.

msimberg commented on include/dlaf/eigensolver/reduction_to_band/impl.h:
```diff
@@ -620,11 +620,11 @@ void updateTrailingPanelWithReflector(const bool has_head, comm::Communicator& c
 }
 
-template <class T>
+template <class T, class CommSender>
```
`CommSender` first.

msimberg commented on include/dlaf/schedulers.h:
```diff
@@ -35,8 +35,10 @@ auto getBackendScheduler() {
+auto getBackendScheduler(const pika::threads::thread_priority priority,
+                         const pika::threads::thread_schedule_hint hint) {
+  namespace ex = pika::execution::experimental;
+  return ex::with_hint(ex::with_priority(getBackendScheduler<backend>(), priority), hint);
```
`prefer` the hint. `cuda_scheduler` doesn't support hints.

msimberg replied:
Removed and needs to be fixed (https://github.com/pika-org/pika/issues/208).

msimberg commented on miniapp/miniapp_band_to_tridiag.cpp:
```diff
@@ -109,6 +109,9 @@ struct BandToTridiagMiniapp {
       if (0 == world.rank() && run_index >= 0)
         std::cout << "[" << run_index << "]" << std::endl;
 
+      pika::scoped_annotation ann1(std::string("miniapp_band_to_tridiag/run_") +
+                                   std::to_string(run_index));
```
Remove.

msimberg commented on spack/packages/dla-future/package.py:
```diff
@@ -38,7 +38,7 @@ class DlaFuture(CMakePackage, CudaPackage):
 
     depends_on("pika cxxstd=17 +mpi")
-    depends_on("pika@0.3.0")
+    depends_on("pika@0.3.0:")
```
Will require 0.4.0.

msimberg commented:
bors try-

rasolca commented:
Note: /project/csstaff/rasolca/DLA-Future/test/unit/common/test_pipeline.cpp doesn't compile.

msimberg commented on test/unit/common/test_pipeline.cpp:
```diff
@@ -65,17 +36,18 @@ TEST(Pipeline, BasicWithSenderAdaptors) {
 
-  EXPECT_FALSE(guard0.is_ready());
-  EXPECT_FALSE(guard1.is_ready());
+  // TODO: What would be a reasonable test here now that guard0 is not a future?
```
This has to be updated with a reasonable new test. Senders don't have `is_ready` in general.

msimberg commented on include/dlaf/schedulers.h:
```diff
@@ -35,8 +35,13 @@ auto getBackendScheduler() {
+                         const pika::threads::thread_schedule_hint) {
+  namespace ex = pika::execution::experimental;
+  using pika::experimental::prefer;
+  // TODO: prefer is broken with with_hint. Try to fix for pika 0.4.0?
+  // return prefer(ex::with_hint, ex::with_priority(getBackendScheduler<backend>(), priority), hint);
```
This needs fixing in any case in pika (see https://github.com/pika-org/pika/issues/208), but I may remove it altogether if we decide that we don't have a use for scheduling hints at the moment.

msimberg commented:
bors try

msimberg commented:
bors try

msimberg commented on CMakeLists.txt:
```diff
@@ -94,7 +94,7 @@ else()
 
 # ----- pika
-find_package(pika 0.3.0 REQUIRED EXACT)
+find_package(pika REQUIRED)
```
Waiting for #538.

msimberg commented:
This is not ready to merge because of the pika 0.4.0 requirement, but I think this is ready to be reviewed. I've removed the various "bonus changes" that I added for testing `band_to_tridiag`. Likely needs some coordination with the changes @rasolca has made to `band_to_tridiag`.

albestro commented on include/dlaf/eigensolver/band_to_tridiag/mc.h:
```diff
@@ -471,7 +473,7 @@ TridiagResult<T, Device::CPU> BandToTridiag<Backend::MC, D, T>::call_L(
           ex::then(store_tau_v) | ex::start_detached();
       deps[step] = dlaf::internal::whenAllLift(w_pipeline(), deps[dep_index]) |
-                   dlaf::internal::transform(policy_hp, cont_sweep) | ex::make_future();
+                   dlaf::internal::transform(policy_hp, cont_sweep) | ex::ensure_started() | ex::split();
```
not sure about that. @rasolca knows the algorithm and underlying mechanisms better than me for sure.

rasolca replied:
They should all be used, unless a bug is hiding somewhere...

msimberg replied:
I've removed the additional `ensure_started` and all seems to be ok.

albestro reviewed (approved):
LGTM.

Waiting for work synchronization on `band_to_tridiag` before reviewing it.

Just one curiosity: I see you prefer to use `CommSender&&` (universal reference) instead of getting it by value as `CommSender`. Is there any reason in particular?

albestro commented:
> > Just one curiosity: I see you prefer to use `CommSender&&` (universal reference) instead of getting it by value as `CommSender`. Is there any reason in particular?
> 
> Mostly because I can and it's generally the right thing to do because it means you usually don't have to worry about unnecessary copies or move construction. I think in this case it makes almost no difference. I may not be thining about this practically enough though. Do you have a reason to prefer taking it by value?

This is the [guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-conventional) I try to follow, because I use to forget all the fancy things compilers can do, so I just try to stick to this because it seems a well thought guideline ([here where the guideline were presented/detailed](https://www.youtube.com/watch?v=xnqTKD8uD64&t=3317s)). It was just a chance to think about it as a possible optimization.

msimberg commented:
> This is the [guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-conventional) I try to follow, because I use to forget all the fancy things compilers can do, so I just try to stick to this because it seems a well thought guideline ([here where the guideline were presented/detailed](https://www.youtube.com/watch?v=xnqTKD8uD64&t=3317s)). It was just a chance to think about it as a possible optimization.

I agree with that guideline. However (though I'm not sure the guideline agrees with me on this, it's not explicitly stated as far as I can tell), that mainly applies to concrete types. There's also https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f19-for-forward-parameters-pass-by-tp-and-only-stdforward-the-parameter. As already mentioned we know that the type is always a future and it's cheaply moved, but it may become a different type with the sender refactorings. It may become different types at different callsites. It may become a sender of a sender (of a sender (of a sender)) which may already be bigger type. All this is to say that when you have a concrete type the benefit of having multiple overloads for lvalues and rvalues is negligible and having a single by-value overload is simple and effective. On the other hand I see the cost of doing a forwarding reference+`forward` also as negligible in this situation because it's a template. But :shrug:

msimberg commented:
bors try
